src/gcode.py

```diff
@@ -59,116 +59,163 @@ def xyz(self, rot):
-
-
 class Position:
-    def __init__(self, X=None, Y=None, Z=None, U=None, V=None):
+    def __init__(self, X=None, Y=None, Z=None, U=None, V=None, E=None):
         self.X = X
         self.Y = Y
         self.Z = Z
         self.U = U
         self.V = V
+        self.E = E
 
     def apply(self, pos):
-        for key in "XYZUV":
+        for key in "XYZUVE":
             val = getattr(pos, key)
             if val is not None:
                 setattr(self, key, val)
 
+    def getCopy(self):
+        return Position(self.X, self.Y, self.Z, self.U, self.V, self.E)
 
-class Printer:
-    def __init__(self):
-        self.currPos = Position(0, 0, 0, 0, 0)
-        self.prevPos = Position(0, 0, 0, 0, 0)
-
-    def updatePos(self, args):
-        # update previous position
-        self.prevPos.apply(self.currPos)
 
+class Printer:
+    def __init__(self, s):
+        self.currPos = Position(0, 0, 0, 0, 0, 0)
+        self.prevPos = Position(0, 0, 0, 0, 0, 0)
+
+        self.rotationPoint = np.array([s.hardware.rotation_center_x, s.hardware.rotation_center_y, s.hardware.rotation_center_z])
+        self.cone_axis = rotation_matrix([1, 0, 0], 0).dot([0, 0, 1])
+
+        self.path = []
+        self.layer = []
+        self.layers = []
+        self.rotations = []
+        self.rotations.append(Rotation(0, 0))
+        self.lays2rots = []
+        self.abs_pos = True  # absolute positioning
+
+        # add an empty layer in the beginning
+        self.layers.append(self.layer)
+        self.lays2rots.append(len(self.rotations) - 1)
+
+    def parseArgs(self, args):
         # convert text args to values
         res = {}
         for arg in args:
             if len(arg) == 0:
                 continue
             key, val = arg[0], arg[1:]
-            if key in "XYZUV":
+            if key in "XYZUVE":
                 res[key] = float(val)
             elif key == ";":
                 break
 
-        # apply values to the printer position
+        return res
+
+    def setAbsPos(self, args):
+        res = self.parseArgs(args)
+
+        for key, val in res.items():
+            setattr(self.currPos, key, val)
+
+    def setRelPos(self, args):
+        res = self.parseArgs(args)
+
         for key, val in res.items():
+            val += getattr(self.currPos, key)
             setattr(self.currPos, key, val)
 
+    def updatePos(self, args):
+        # update previous position
+        self.prevPos.apply(self.currPos)
+
+        # apply values to the printer position
+        if self.abs_pos:
+            self.setAbsPos(args)
+        else:
+            self.setRelPos(args)
+
+        self.dX = self.prevPos.X - self.currPos.X
+        self.dY = self.prevPos.Y - self.currPos.Y
+        self.dZ = self.prevPos.Z - self.currPos.Z
         self.dU = self.prevPos.U - self.currPos.U
+        self.dE = self.prevPos.E - self.currPos.E
+        noMove = self.dX == 0 and self.dY == 0 and self.dZ == 0 and self.dU == 0
```

Comment:
almostZero?)

Reply:
no,I think not needed here, we should behave like a real printer, it considers even a very slow movement as movement

Reply:
I think goosli should care that two sequent lines in gcode differ enaugh